libs/elasticsearch-core/src/test/java/org/elasticsearch/core/internal/io/IOUtilsTests.java

```diff
@@ -0,0 +1,174 @@
+            final Function<Closeable[], T> function, final CheckedConsumer<T, IOException> close) throws IOException {
+        final int numberOfCloseables = randomIntBetween(1, 8);
+        final Closeable[] closeables = new Closeable[numberOfCloseables];
+        final List<Integer> indexesThatThrow = new ArrayList<>(numberOfCloseables);
+        for (int i = 0; i < numberOfCloseables - 1; i++) {
+            final Closeable closeable = mock(Closeable.class);
+            if (randomBoolean()) {
+                indexesThatThrow.add(i);
+                doThrow(new IOException(Integer.toString(i))).when(closeable).close();
+            }
+            closeables[i] = closeable;
+        }
+
+        // ensure that at least one always throws
+        final Closeable closeable = mock(Closeable.class);
+        if (indexesThatThrow.isEmpty() || randomBoolean()) {
+            indexesThatThrow.add(numberOfCloseables - 1);
+            doThrow(new IOException(Integer.toString(numberOfCloseables - 1))).when(closeable).close();
+        }
+        closeables[numberOfCloseables - 1] = closeable;
+
+        final IOException e = expectThrows(IOException.class, () -> close.accept(function.apply(closeables)));
+        assertThat(e.getMessage(), equalTo(Integer.toString(indexesThatThrow.get(0))));
+        assertThat(e.getSuppressed(), arrayWithSize(indexesThatThrow.size() - 1));
+        for (int i = 1; i < indexesThatThrow.size(); i++) {
+            assertNotNull(e.getSuppressed()[i - 1]);
+            assertThat(e.getSuppressed()[i - 1].getMessage(), equalTo(Integer.toString(indexesThatThrow.get(i))));
+        }
+    }
+
+    public void testDeleteFilesIgnoringExceptionsArray() throws IOException {
+        runDeleteFilesIgnoringExceptionsTest(Function.identity(), IOUtils::deleteFilesIgnoringExceptions);
+    }
+
+    public void testDeleteFilesIgnoringExceptionsIterable() throws IOException {
+        runDeleteFilesIgnoringExceptionsTest(Arrays::asList, IOUtils::deleteFilesIgnoringExceptions);
+    }
+
+    private <T> void runDeleteFilesIgnoringExceptionsTest(
+            final Function<Path[], T> function, CheckedConsumer<T, IOException> deleteFilesIgnoringExceptions) throws IOException {
+        final int numberOfFiles = randomIntBetween(0, 7);
+        final Path[] files = new Path[numberOfFiles];
+        for (int i = 0; i < numberOfFiles; i++) {
+            if (randomBoolean()) {
+                files[i] = createTempFile();
+            } else {
+                final Path temporary = createTempFile();
+                files[i] = PathUtils.get(temporary.toString(), randomAlphaOfLength(8));
+                Files.delete(temporary);
+            }
+        }
+        deleteFilesIgnoringExceptions.accept(function.apply(files));
+        for (int i = 0; i < numberOfFiles; i++) {
+            assertFalse(files[i].toString(), Files.exists(files[i]));
+        }
+    }
+
+    public void testRm() throws IOException {
```

Comment:
I am not sure how to test the exceptional case of `IOUtils#rm`, it seems difficult to force an I/O exception while deleting a file; I am open to suggestions or we leave it uncovered (as the Lucene version is uncovered too).

Reply:
Lucene has a FilterFileSystemProvider which allows mocking of various file system calls (for example the VirusCheckingFS class in the lucene test-framework which throws AccessDeniedExceptions on delete)

Reply:
Thanks @romseygeek; I pushed a test in 523e36b61714e18de756327c829b5fb5b3bbfad5.